tools/qlog/src/helpers.rs

```diff
@@ -0,0 +1,841 @@
+                stream_id,
+                offset,
+                from,
+                to,
+                raw,
+            },
+        }
+    }
+
+    pub fn h3_data_moved_min(stream_id: String) -> Self {
+        Event::h3_data_moved(stream_id, None, None, None, None)
+    }
+
+    /// Returns:
+    /// * `EventType`=`Http3EventType::PushResolved`
+    /// * `EventData`=`H3PushResolved`.
+    pub fn h3_push_resolved(
+        push_id: Option<String>, stream_id: Option<String>,
+        decision: Option<H3PushDecision>,
+    ) -> Self {
+        Event {
+            ty: EventType::Http3EventType(Http3EventType::PushResolved),
+            data: EventData::H3PushResolved {
+                push_id,
+                stream_id,
+                decision,
+            },
+        }
+    }
+
+    pub fn h3_push_resolved_min() -> Self {
+        Event::h3_push_resolved(None, None, None)
+    }
+
+    // QPACK events.
+
+    /// Returns:
+    /// * `EventType`=`QpackEventType::StateUpdated`
+    /// * `EventData`=`QpackStateUpdated`.
+    pub fn qpack_state_updated(
+        owner: Option<QpackOwner>, dynamic_table_capacity: Option<u64>,
+        dynamic_table_size: Option<u64>, known_received_count: Option<u64>,
+        current_insert_count: Option<u64>,
+    ) -> Self {
+        Event {
+            ty: EventType::QpackEventType(QpackEventType::StateUpdated),
+            data: EventData::QpackStateUpdated {
+                owner,
+                dynamic_table_capacity,
+                dynamic_table_size,
+                known_received_count,
+                current_insert_count,
+            },
+        }
+    }
+
+    pub fn qpack_state_updated_min() -> Self {
+        Event::qpack_state_updated(None, None, None, None, None)
+    }
+
+    /// Returns:
+    /// * `EventType`=`QpackEventType::StreamStateUpdated`
+    /// * `EventData`=`QpackStreamStateUpdated`.
+    pub fn qpack_stream_state_updated(
+        stream_id: String, state: QpackStreamState,
+    ) -> Self {
+        Event {
+            ty: EventType::QpackEventType(QpackEventType::StreamStateUpdated),
+            data: EventData::QpackStreamStateUpdated { stream_id, state },
+        }
+    }
+
+    /// Returns:
+    /// * `EventType`=`QpackEventType::DynamicTableUpdated`
+    /// * `EventData`=`QpackDynamicTableUpdated`.
+    pub fn qpack_dynamic_table_updated(
+        update_type: QpackUpdateType, entries: Vec<QpackDynamicTableEntry>,
+    ) -> Self {
+        Event {
+            ty: EventType::QpackEventType(QpackEventType::DynamicTableUpdated),
+            data: EventData::QpackDynamicTableUpdated {
+                update_type,
+                entries,
+            },
+        }
+    }
+
+    /// Returns:
+    /// * `EventType`=`QpackEventType::HeadersEncoded`
+    /// * `EventData`=`QpackHeadersEncoded`.
+    pub fn qpack_headers_encoded(
+        stream_id: Option<String>, headers: Option<HttpHeader>,
+        block_prefix: QpackHeaderBlockPrefix,
```

Comment:
I am not a sophisticated rust programmer, but I cannot determine how a user of this function can build a `QpackHeaderBlockPrefix`. While the struct itself is public, none of the fields are. Therefore, it seems impossible to create such an object outside the package and pass it as a parameter here. The same seems to apply to `QpackDynamicTableEntry`. Please forgive me if this is something that any reasonable rust programmer should know and I am saying something stupid. 

No matter what, thank you for creating this PR. I am looking forward to all these helper methods :-)

Reply:
Hehe, this is not sophistication but rather a bug caused by not testing out the code from external modules.

We should make the fields public too, which will fix the problem